From afd02b93fc02a74aa0d210eeca67c39957f92482 Mon Sep 17 00:00:00 2001
From: Lionel Debieve <lionel.debieve@foss.st.com>
Date: Wed, 2 Feb 2022 21:54:53 +0100
Subject: [PATCH 0308/1695] plat-stm32mp1: retrieve chip id from syscfg for
 STM32MP13

Chip ID is read from SYSCFG. Add the associated read
function and new CHIP IDs.

Use the chip id to dynamically detect the crypto hardware
support and the second cpu core.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git c9243a4b109f39ad263e8e7e26bf57a1faf3b002 plat-stm32mp1: retrieve chip id from syscfg for STM32MP13]
Signed-off-by: Lionel Debieve <lionel.debieve@st.com>
Change-Id: Ic637031863bf4dba785c86c49dd762c80720b282
---
 .../plat-stm32mp1/drivers/stm32mp1_syscfg.c   |  17 +++
 core/arch/arm/plat-stm32mp1/main.c            | 108 ++++++++++++++++++
 core/arch/arm/plat-stm32mp1/stm32_util.h      |  10 ++
 3 files changed, 135 insertions(+)

diff --git a/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_syscfg.c b/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_syscfg.c
index 1b2ecbcff..edf32c42d 100644
--- a/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_syscfg.c
+++ b/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_syscfg.c
@@ -24,6 +24,7 @@
 #define SYSCFG_CMPSD1CR				U(0x30)
 #define SYSCFG_CMPSD2CR				U(0x40)
 #define SYSCFG_HSLVEN0R				U(0x50)
+#define SYSCFG_IDC				U(0x380)
 #define SYSCFG_IOSIZE				U(0x400)
 
 /*
@@ -61,6 +62,13 @@
 #define SYSCFG_HSLV_MASK			GENMASK_32(15, 0)
 #define SYSCFG_HSLV_KEY				U(0x1018)
 
+/*
+ * SYSCFG_IDC Register
+ */
+#define SYSCFG_IDC_DEV_ID_MASK			GENMASK_32(11, 0)
+#define SYSCFG_IDC_REV_ID_MASK			GENMASK_32(31, 16)
+#define SYSCFG_IDC_REV_ID_SHIFT			U(16)
+
 static vaddr_t get_syscfg_base(void)
 {
 	static struct io_pa_va base = { .pa = SYSCFG_BASE };
@@ -68,6 +76,15 @@ static vaddr_t get_syscfg_base(void)
 	return io_pa_or_va(&base, SYSCFG_IOSIZE);
 }
 
+uint32_t stm32mp_syscfg_get_chip_dev_id(void)
+{
+	if (IS_ENABLED(CFG_STM32MP13))
+		return io_read32(get_syscfg_base() + SYSCFG_IDC) &
+		       SYSCFG_IDC_DEV_ID_MASK;
+
+	return 0;
+}
+
 static void enable_io_compensation(int cmpcr_offset)
 {
 	vaddr_t cmpcr_va = get_syscfg_base() + cmpcr_offset;
diff --git a/core/arch/arm/plat-stm32mp1/main.c b/core/arch/arm/plat-stm32mp1/main.c
index 373eec118..d382ced80 100644
--- a/core/arch/arm/plat-stm32mp1/main.c
+++ b/core/arch/arm/plat-stm32mp1/main.c
@@ -513,3 +513,111 @@ early_init_late(init_debug);
 
 /* Some generic resources need to be unpaged */
 DECLARE_KEEP_PAGER(pinctrl_apply_state);
+
+static int get_chip_dev_id(uint32_t *dev_id)
+{
+#ifdef CFG_STM32MP13
+	*dev_id = stm32mp_syscfg_get_chip_dev_id();
+#else /* assume CFG_STM32MP15 */
+	if (stm32mp1_dbgmcu_get_chip_dev_id(dev_id) != TEE_SUCCESS)
+		return -1;
+#endif
+	return 0;
+}
+
+static int get_part_number(uint32_t *part_nb)
+{
+	static uint32_t part_number;
+	uint32_t dev_id = 0;
+	uint32_t otp = 0;
+	size_t bit_len = 0;
+
+	assert(part_nb);
+
+	if (part_number) {
+		*part_nb = part_number;
+		return 0;
+	}
+
+	if (get_chip_dev_id(&dev_id) < 0)
+		return -1;
+
+	if (stm32_bsec_find_otp_in_nvmem_layout("part_number_otp",
+						&otp, NULL, &bit_len))
+		return -1;
+
+	if (stm32_bsec_read_otp(&part_number, otp))
+		return -1;
+
+	part_number = (part_number & GENMASK_32(bit_len, 0));
+
+	part_number = part_number | (dev_id << 16);
+
+	*part_nb = part_number;
+
+	return 0;
+}
+
+bool stm32mp_supports_hw_cryp(void)
+{
+	uint32_t part_number = 0;
+
+	if (!IS_ENABLED(CFG_STM32_CRYP))
+		return false;
+
+	if (get_part_number(&part_number)) {
+		DMSG("Cannot get part number\n");
+		panic();
+	}
+
+	switch (part_number) {
+#ifdef CFG_STM32MP13
+	case STM32MP135F_PART_NB:
+	case STM32MP135C_PART_NB:
+	case STM32MP133F_PART_NB:
+	case STM32MP133C_PART_NB:
+	case STM32MP131F_PART_NB:
+	case STM32MP131C_PART_NB:
+		return true;
+#endif
+#ifdef CFG_STM32MP15
+	case STM32MP157F_PART_NB:
+	case STM32MP157C_PART_NB:
+	case STM32MP153F_PART_NB:
+	case STM32MP153C_PART_NB:
+	case STM32MP151F_PART_NB:
+	case STM32MP151C_PART_NB:
+		return true;
+#endif
+	default:
+		return false;
+	}
+}
+
+bool stm32mp_supports_second_core(void)
+{
+	uint32_t __maybe_unused part_number = 0;
+
+	if (IS_ENABLED(CFG_STM32MP13))
+		return false;
+
+	if (CFG_TEE_CORE_NB_CORE == 1)
+		return false;
+
+	if (get_part_number(&part_number)) {
+		DMSG("Cannot get part number\n");
+		panic();
+	}
+
+	switch (part_number) {
+#ifdef CFG_STM32MP15
+	case STM32MP151F_PART_NB:
+	case STM32MP151D_PART_NB:
+	case STM32MP151C_PART_NB:
+	case STM32MP151A_PART_NB:
+		return false;
+#endif
+	default:
+		return true;
+	}
+}
diff --git a/core/arch/arm/plat-stm32mp1/stm32_util.h b/core/arch/arm/plat-stm32mp1/stm32_util.h
index 9dbc16613..d2e30e34f 100644
--- a/core/arch/arm/plat-stm32mp1/stm32_util.h
+++ b/core/arch/arm/plat-stm32mp1/stm32_util.h
@@ -12,6 +12,7 @@
 #include <drivers/stm32_bsec.h>
 #include <drivers/stm32mp1_rcc_util.h>
 #include <kernel/panic.h>
+#include <stdbool.h>
 #include <stdint.h>
 #include <tee_api_types.h>
 #include <types_ext.h>
@@ -20,9 +21,18 @@
 TEE_Result stm32mp1_dbgmcu_get_chip_version(uint32_t *chip_version);
 TEE_Result stm32mp1_dbgmcu_get_chip_dev_id(uint32_t *chip_dev_id);
 
+/*  Crypto HW support */
+bool stm32mp_supports_hw_cryp(void);
+
+/*  Second core support */
+bool stm32mp_supports_second_core(void);
+
 /* Backup registers and RAM utils */
 vaddr_t stm32mp_bkpreg(unsigned int idx);
 
+/* Get device ID from SYSCFG registers */
+uint32_t stm32mp_syscfg_get_chip_dev_id(void);
+
 /* Platform util for the GIC */
 vaddr_t get_gicd_base(void);
 
-- 
2.47.2

